File: stack/daily-temperatures.js
// 739. Daily Temperatures
// https://leetcode.com/problems/daily-temperatures/description/

function dailyTemperatures(temperatures) {
  // create an array to store the number of days to wait for a warmer temperature
  const waitingDays = new Array(temperatures.length).fill(0);

  // create a stack to store the indices of the temperatures array
  const stack = [];

  // iterate through the temperatures array
  for (let i = 0; i < temperatures.length; i++) {
    // while the stack is not empty and the current temperature is warmer than the temperature at the top of the stack
    while (
      stack.length > 0 &&
      temperatures[i] > temperatures[stack[stack.length - 1]]
    ) {
      // pop the top element from the stack
      const index = stack.pop();
      // update the number of days to wait for a warmer temperature at the popped index
      waitingDays[index] = i - index;
    }
    // push the current index to the stack
    stack.push(i);
  }

  // return the array of waiting days
  return waitingDays;
}

File: daily-temperatures.java
// https://leetcode.com/problems/daily-temperatures/description/

// Java
class Solution {
    public int[] dailyTemperatures(int[] temps) {
        int[] res = new int[temps.length];
        Deque<Integer> d = new ArrayDeque<>();

        for (int i = temps.length - 1; i >= 0; i--) {
            while (!d.isEmpty() && temps[d.peekLast()] <= temps[i]) d.pollLast();
            res[i] = d.isEmpty() ? 0 : d.peekLast() - i;
            d.offerLast(i);
        }

        return res;
    }
}

// monotonic decreasing stack - temps on top must be > than the temo below
// The stack stores indices of temperatures, not the temperatures themselves.
// We iterate from the end of the array to the beginning, since we're looking for the "next" warmer day (which lies ahead in time).
// For each day:
    // We remove indices from the stack as long as their corresponding temperatures are less than or equal to the current day's temperature. These can't help us anymore.
    // If the stack is not empty, the top of the stack points to the next warmer day. The difference in indices tells us how many days to wait.
    // If the stack is empty, there's no warmer day in the future, so we record 0.
// After processing each day, we push its index onto the stack to possibly help future (earlier) days.
